build/builder.go

```diff
@@ -505,15 +505,44 @@ func (o *Options) FindAllShards() []string {
 
 // IgnoreSizeMax determines whether the max size should be ignored.
 func (o *Options) IgnoreSizeMax(name string) bool {
+	matched := false
 	for _, pattern := range o.LargeFiles {
 		pattern = strings.TrimSpace(pattern)
-		m, _ := doublestar.PathMatch(pattern, name)
+		negated, validatedPattern := checkIsNegatePattern(pattern)
+
+		m, _ := doublestar.PathMatch(validatedPattern, name)
 		if m {
-			return true
+			// match against a negated LargeFiles pattern overrides all non-exclude matches
+			if negated {
+				return false
+			}
+			matched = true
 		}
 	}
 
-	return false
+	return matched
+}
+
+func checkIsNegatePattern(pattern string) (bool, string) {
+	negate := "!"
+
+	if len(pattern) < len(negate) {
+		return false, pattern
+	}
+
+	// if negated then strip prefix meta characters representing negated filter pattern
+	if strings.Index(pattern, negate) == 0 {
+		return true, pattern[len(negate):]
+	}
+
+	escape := "\\"
```

Comment:
I could use Replacer here but since we probably want to keep the amount of parsing of our own custom syntax to a minimum I decided to just keep it simple.